# Customer Churn Prediction System

## Phase 2: Data Collection and Understanding

This notebook performs initial data understanding of the Telco Customer Churn dataset, including dataset structure analysis, feature understanding, target variable identification, missing value detection, and data type analysis.


## Activity 1 - Load Dataset
Load the Telco Customer Churn dataset into a pandas DataFrame and preview the first 5 records to confirm it loaded correctly.

In [3]:
import pandas as pd

# Load dataset
df = pd.read_csv('../data/telco_customer_churn.csv')

# Preview first 5 records
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


**Observation:** The dataset loaded successfully. The first 5 rows confirm that column names, values, and formatting are as expected for a telecom customer dataset.

## Activity 2 - Review Dataset Structure
Inspect the shape, column names, and general information of the dataset to understand its size and composition.

In [3]:
print("=== Dataset Shape ===")
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

print("\n=== Column Names ===")
print(df.columns.tolist())

print("\n=== Dataset Information ===")
df.info()

=== Dataset Shape ===
Rows: 7043
Columns: 21

=== Column Names ===
['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn']

=== Dataset Information ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8 

**Observation:** The dataset has 7,043 rows and 21 columns. All column names are meaningful business terms. The target variable is `Churn`. The `customerID` column is an identifier and will be dropped before model training.

## Activity 3 - Understand Feature Definitions
Generate numerical and categorical feature summaries to understand value distributions, ranges, and unique value counts for each feature.

In [4]:
print("=== Numerical Feature Summary ===")
display(df.describe())

print("\n=== Categorical Feature Summary ===")
display(df.describe(include='object'))

=== Numerical Feature Summary ===


,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000



=== Categorical Feature Summary ===


,customerID,gender,Partner,Dependents,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,TotalCharges,Churn
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043
unique,7043,2,2,2,2,3,3,3,3,3,3,3,3,3,2,4,6531,2
top,3186-AJIEK,Male,No,No,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,20.2,No
freq,1,3555,3641,4933,6361,3390,3096,3498,3088,3095,3473,2810,2785,3875,4171,2365,11,5174


**Observation:** There are 3 numerical features - `SeniorCitizen`, `tenure`, and `MonthlyCharges`. Tenure ranges from 0 to 72 months. MonthlyCharges ranges from `$18.25` to `$118.75`. There are 18 categorical (object) features. Note: `TotalCharges` is missing from the numerical summary because it is incorrectly stored as a string - this will be fixed during preprocessing.

## Activity 4 - Identify Target Variable
Identify the target variable `Churn`, examine its unique values, class distribution, and class percentages to understand the prediction task.

In [5]:
print("=== Target Variable ===")
print("Churn")

print("\n=== Unique Values ===")
print(df['Churn'].unique())

print("\n=== Churn Class Counts ===")
print(df['Churn'].value_counts())

print("\n=== Churn Class Percentages ===")
print((df['Churn'].value_counts(normalize=True) * 100).round(2))

=== Target Variable ===
Churn

=== Unique Values ===
['No' 'Yes']

=== Churn Class Counts ===
Churn
No     5174
Yes    1869
Name: count, dtype: int64

=== Churn Class Percentages ===
Churn
No     73.46
Yes    26.54
Name: proportion, dtype: float64


**Observation:** The target variable `Churn` has exactly 2 values - `Yes` (churned) and `No` (retained), confirming this is a binary classification problem. The dataset is moderately imbalanced: 73.46% No and 26.54% Yes. This means Recall will be prioritised over Accuracy as the primary evaluation metric during model training.

## Activity 5 - Detect Missing Values
Check for both formal null (NaN) values and hidden blank whitespace entries, which pandas does not detect as null by default.

In [6]:
print("=== Missing Values Per Column ===")
display(df.isnull().sum())

print("\n=== Total Missing Values ===")
print(df.isnull().sum().sum())

print("\n=== Blank Values in Dataset ===")
display((df == " ").sum())

print("\n=== Blank Values in TotalCharges ===")
print((df["TotalCharges"] == " ").sum())

=== Missing Values Per Column ===


customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


=== Total Missing Values ===
0

=== Blank Values in Dataset ===


customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64


=== Blank Values in TotalCharges ===
11


**Observation:** No formal null (NaN) values exist in any column. However, `TotalCharges` contains 11 blank whitespace entries (' ') that are not detected by `isnull()`. These must be handled separately during preprocessing using `pd.to_numeric(df['TotalCharges'], errors='coerce')`, which converts blanks to NaN, followed by median imputation.

## Activity 6 - Analyze Data Types
Inspect the data type of each column, count how many columns belong to each type, and separate features into numerical and categorical groups.

In [7]:
print("=== Data Types ===")
display(df.dtypes)

print("\n=== Data Type Counts ===")
display(df.dtypes.value_counts())

numerical_features = df.select_dtypes(
    include=['int64', 'float64']
).columns

categorical_features = df.select_dtypes(
    include=['object']
).columns

print("\n=== Numerical Features ===")
print(list(numerical_features))

print("\n=== Categorical Features ===")
print(list(categorical_features))

=== Data Types ===


customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object


=== Data Type Counts ===


object     18
int64       2
float64     1
Name: count, dtype: int64


=== Numerical Features ===
['SeniorCitizen', 'tenure', 'MonthlyCharges']

=== Categorical Features ===
['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'TotalCharges', 'Churn']


**Observation:** pandas detects 3 numerical features (SeniorCitizen, tenure, MonthlyCharges) and 18 object features. `TotalCharges` incorrectly appears in the categorical list due to its data type error. After fixing it with `pd.to_numeric()`, it will move to the numerical group, bringing the correct numerical count to 4.

## Feature Summary Table
A combined summary showing the data type, missing value count, and number of unique values for all 21 features.

In [4]:
feature_summary = pd.DataFrame({
    'Data Type': df.dtypes,
    'Missing Values': df.isnull().sum(),
    'Unique Values': df.nunique()
})

feature_summary

,Data Type,Missing Values,Unique Values
customerID,object,0,7043
gender,object,0,2
SeniorCitizen,int64,0,2
Partner,object,0,2
Dependents,object,0,2
tenure,int64,0,73
PhoneService,object,0,2
MultipleLines,object,0,3
InternetService,object,0,3
OnlineSecurity,object,0,3


**Observation:** The feature summary confirms that only `TotalCharges` has a data quality issue (wrong data type + 11 blank entries). All other 20 columns are complete with no missing values. `customerID` has 7,043 unique values (one per row), confirming it is a row identifier that must be dropped before training.

## Key Findings

- The dataset contains 7,043 customer records and 21 features.
- `Churn` is the target variable and represents a binary classification problem (Yes / No).
- The dataset contains both numerical features (SeniorCitizen, tenure, MonthlyCharges) and categorical features (18 object columns).
- No formal null (NaN) values are present in any column.
- `TotalCharges` contains 11 blank whitespace entries and is incorrectly stored as object type - both issues must be fixed during preprocessing.
- The dataset is moderately class-imbalanced: 73.46% No Churn vs 26.54% Churn - Recall will be the primary evaluation metric.
- The dataset is suitable for customer churn prediction using machine learning techniques.
- Additional preprocessing and feature engineering will be performed before model development.